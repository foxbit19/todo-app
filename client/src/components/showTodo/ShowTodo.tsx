import React from 'react'
import { Button, Dialog, DialogActions, DialogContent, DialogContentText, DialogTitle, Slide } from '@mui/material'
import Item from '../../models/item'
import UpdateIcon from '@mui/icons-material/Update'
import DoneIcon from '@mui/icons-material/Done'
import { TransitionProps } from '@mui/material/transitions'

interface Props {
    open: boolean
    item: Item
    onUpdateClick?: (item: Item) => void
    onDeleteClick?: (item: Item) => void
    onClose?: () => void
}

const Transition = React.forwardRef(function Transition(
    props: TransitionProps & {
        children: React.ReactElement<any, any>;
    },
    ref: React.Ref<unknown>,
) {
    return <Slide direction="up" ref={ref} {...props} />;
});

const ShowTodo = (props: Props) => {

    const handleUpdateClick = () => {
        if (props.onUpdateClick) {
            props.onUpdateClick(props.item)
        }
    }

    const handleDeleteClick = () => {
        if (props.onDeleteClick) {
            props.onDeleteClick(props.item)
        }
    }

    return <Dialog open={props.open} onClose={props.onClose} TransitionComponent={Transition}>
        <DialogTitle>Item details</DialogTitle>
        <DialogContent>
            <DialogContentText>
                {props.item.description}
            </DialogContentText>
        </DialogContent>
        <DialogActions>
            <Button data-testid='update_button' onClick={handleUpdateClick} startIcon={<UpdateIcon />}>Update</Button>
            <Button data-testid='delete_button' onClick={handleDeleteClick} startIcon={<DoneIcon />}>Complete</Button>
        </DialogActions>
    </Dialog>
}

export default ShowTodo