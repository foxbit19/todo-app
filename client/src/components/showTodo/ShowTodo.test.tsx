import React from 'react';
import { render, screen } from '@testing-library/react';
import ShowTodo from './ShowTodo';
import Item from '../../models/item';
import sinon from 'sinon';

describe('show todo', () => {
    let item: Item

    beforeEach(() => {
        item = {
            id: 1,
            description: 'this is my item',
            order: 2
        }
    })

    const clickButton = (testId: string) => {
        screen.getByTestId(testId).click()
    }

    test('it matches snapshot', () => {
        const view = render(<ShowTodo item={item} />);
        expect(view).toMatchSnapshot();
    });

    test('it shows the description of the provided item', () => {
        render(<ShowTodo item={item} />)

        expect(screen.getByText(item.description)).toBeInTheDocument()
    })

    test('it shows the "item details" text', () => {
        render(<ShowTodo item={item} />)

        expect(screen.getByText('Item details')).toBeInTheDocument()
    })

    test('it launch an event when update button is clicked', () => {
        const updateHandler = sinon.spy();

        render(<ShowTodo item={item} onUpdateClick={updateHandler} />)
        clickButton('update_button')
        expect(updateHandler.called).toBeTruthy()
    })

    test('it launch an event when delete button is clicked', () => {
        const deleteHandler = sinon.spy();

        render(<ShowTodo item={item} onDeleteClick={deleteHandler} />)
        clickButton('delete_button')
        expect(deleteHandler.called).toBeTruthy()
    })

    test('it launch update event providing the item as argument', () => {
        const updateHandler = sinon.spy();

        render(<ShowTodo item={item} onUpdateClick={updateHandler} />)
        clickButton('update_button')
        expect(updateHandler.calledWith(item)).toBeTruthy()
    })

    test('it launch delete event providing the item as argument', () => {
        const deleteHandler = sinon.spy();

        render(<ShowTodo item={item} onDeleteClick={deleteHandler} />)
        clickButton('delete_button')
        expect(deleteHandler.calledWith(item)).toBeTruthy()
    })
})