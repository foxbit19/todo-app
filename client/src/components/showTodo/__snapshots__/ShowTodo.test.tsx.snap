// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show todo it matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
        role="presentation"
        style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          aria-labelledby=":r0:"
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1t1j96h-MuiPaper-root-MuiDialog-paper"
          role="dialog"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
            id=":r0:"
          >
            Item details
          </h2>
          <div
            class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
          >
            <p
              class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiDialogContentText-root css-qfso29-MuiTypography-root-MuiDialogContentText-root"
            >
              this is my item
            </p>
          </div>
          <div
            class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              data-testid="update_button"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="UpdateIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M21 10.12h-6.78l2.74-2.82c-2.73-2.7-7.15-2.8-9.88-.1-2.73 2.71-2.73 7.08 0 9.79s7.15 2.71 9.88 0C18.32 15.65 19 14.08 19 12.1h2c0 1.98-.88 4.55-2.64 6.29-3.51 3.48-9.21 3.48-12.72 0-3.5-3.47-3.53-9.11-.02-12.58s9.14-3.47 12.65 0L21 3v7.12zM12.5 8v4.25l3.5 2.08-.72 1.21L11 13V8h1.5z"
                  />
                </svg>
              </span>
              Update
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              data-testid="delete_button"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DoneIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
                  />
                </svg>
              </span>
              Complete
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
